latch/ldata/path.py

```diff
@@ -0,0 +1,271 @@
+import os
+import re
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Generator, Optional, Type, Union
+
+import gql
+from flytekit import (
+    Blob,
+    BlobMetadata,
+    BlobType,
+    FlyteContext,
+    Literal,
+    LiteralType,
+    Scalar,
+)
+from flytekit.extend import TypeEngine, TypeTransformer
+from latch_sdk_gql.execute import execute
+
+from latch.ldata.transfer.node import LDataNodeType
+from latch.ldata.transfer.progress import Progress
+from latch_cli.utils import urljoins
+
+from .transfer.download import _download
+from .transfer.remote_copy import _remote_copy
+from .transfer.upload import _upload
+
+node_id_regex = re.compile(r"^latch://(?P<id>[0-9]+)\.node$")
+
+
+dir_types = {
+    LDataNodeType.dir,
+    LDataNodeType.account_root,
+    LDataNodeType.mount,
+}
+
+
+@dataclass
+class _Cache:
+    """Internal cache class to organize information for a `LPath`."""
+
+    path: Optional[str] = None
+    node_id: Optional[str] = None
+    name: Optional[str] = None
+    type: Optional[LDataNodeType] = None
+    size: Optional[int] = None
+    content_type: Optional[str] = None
+
+
+download_idx = 0
+
+
+@dataclass(frozen=True)
+class LPath:
+
+    _cache: _Cache = field(
+        default_factory=lambda: _Cache(),
+        init=False,
+        repr=False,
+        hash=False,
+        compare=False,
+    )
+
+    path: str
+
+    def __post_init__(self):
+        if isinstance(self.path, LPath):
```

Comment:
yeah I added this because i thought it would be a common error. for example if someone tries to do this

```
x = LPath("latch://...")
y = LPath(x / "test.txt")
```
this will cause errors that might confuse the developer, feel like its safer to just have the check